File: android/src/main/java/co/alephnull/reactnative/silicon/Payload.kt
package co.alephnull.reactnative.silicon

import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class BridgePayloadRecord : Record {
    @Field var text: String? = null
    @Field var bytes: ByteArray? = null
}

sealed interface PayloadType
@JvmInline
value class PayloadText(val text: String) : PayloadType
@JvmInline
value class PayloadByteArr(val arr: ByteArray) : PayloadType

/**
 * Extension function to safely cross from BridgePayloadRecord to PayloadType
 */
fun BridgePayloadRecord.toPayloadType(): PayloadType {
    return when {
        text != null -> PayloadText(text as String)
        bytes != null -> PayloadByteArr(bytes as ByteArray)
        else -> throw IllegalArgumentException("Payload record must contain either 'text' or 'bytes'.")
    }
}

File: ios/Payload.swift
import ExpoModulesCore

struct BridgePayloadRecord: Record {
    @Field var text: String?
    @Field var bytes: Data? // Expo maps JS Uint8Array to Swift Data
}

enum PayloadType {
    case text(String)
    case byteArr(Data)
}

extension BridgePayloadRecord {
    func toPayloadType() throws -> PayloadType {
        if let text = self.text {
            return .text(text)
        } else if let bytes = self.bytes {
            return .byteArr(bytes)
        } else {
            throw NSError(
                domain: "Silicon",
                code: 0,
                userInfo: [NSLocalizedDescriptionKey: "Payload record must contain either text or bytes."]
            )
        }
    }
}
